Do not allow the daemon to start when MaxDownloadAttempts <=0 fixes regression #44921

psaintlaurent opened the pull request:
**- What I did**

Fixed regression that allows the daemon to be started with max-download-attempts=0

**- How I did it**
Modified the config.go

**- How to verify it**
Start a moby container 
`make BIND_DIR=. shell`

**Start the daemon with --max-download-attempts=0**
```
./hack/make.sh binary
make install
/usr/local/bin/dockerd --max-download-attempts=0
```
<img width="1346" alt="Screenshot 2024-06-25 at 11 14 03 AM" src="https://github.com/moby/moby/assets/347096/ff00f957-5bc9-4f29-a633-6eef4a644ae8">

```
mv /etc/docker/daemon.json /etc/docker/daemon.json.bak
/usr/local/bin/dockerd --max-download-attempts=0
```

**Move the daemon.json file temporarily start the daemon again with --max-download-attempts=0**
<img width="1347" alt="Screenshot 2024-06-25 at 11 17 13 AM" src="https://github.com/moby/moby/assets/347096/7c26b38e-641f-4a77-873d-92acc38db9fd">


**- Description for the changelog**

```
Do not allow the daemon to start when MaxDownloadAttempts <=0 fixes regression #44921
```

psaintlaurent committed:
Do not allow the daemon to start when MaxDownloadAttempts <=0

Signed-off-by: plaurent <patrick@saint-laurent.us>

dperny reviewed (changes requested):
There's probably some more that is needed for this fix, or a fix in a different location, where we need to ensure that the default value of 5 is correctly set before we hit validation.

dperny commented:
I'm not certain what the fix is, but this isn't quite right.

psaintlaurent commented on cmd/dockerd/daemon_test.go:
```diff
@@ -146,7 +146,7 @@ func TestLoadDaemonCliConfigWithoutTLSVerify(t *testing.T) {
 
 func TestLoadDaemonCliConfigWithLogLevel(t *testing.T) {
-	tempFile := fs.NewFile(t, "config", fs.WithContent(`{"log-level": "warn"}`))
+	tempFile := fs.NewFile(t, "config", fs.WithContent(`{"log-level": "warn", "max-download-attempts": 5}`))
```
ok great

psaintlaurent commented on REBUILD_TEST:
I wanted to trigger a rebuild of the pipeline but I don't see any existing options to do so.

dperny replied:
You can trigger a rebuild by doing a force-push. You should feel free and comfortable to force-push to your PR at any time. Force-push is an expected part of the workflow.

I am in the habit of using `--force-with-lease` when I force-push, because it stops you from accidentally clobbering anything that changed on the remote branch by accident.

dperny commented:
Put another way:

1. The expected behavior is that if the value of `max-download-attempts` is `0`, the daemon should fail to start.
2. The expected behavior is that if the user does not set a value of `max-download-attempts`, then the value of `max-download-attempts` should be `5`.
3. Because of 1 and 2, the daemon should start if no value is passed for `max-download-attempts`.

dperny reviewed (approved):
This looks good to me.

dperny commented:
... except for the fact that there's this extra file, that needs to go, obviously.

corhere reviewed (changes requested):
This change behaves incorrectly.
```console
root@56e9443a7364:/go/src/github.com/docker/docker# cat /etc/docker/daemon.json
{
	"max-download-attempts": 7,
	"runtimes": {
		"crun": {
			"path": "/usr/local/bin/crun"
		}
	}
}
root@56e9443a7364:/go/src/github.com/docker/docker# bundles/dynbinary-daemon/dockerd --userland-proxy=false -D 2>&1| grep 'Max Download Attempts'
time="2024-06-28T17:59:13.584092212Z" level=debug msg="Max Download Attempts: 5"
^C
```

It's unfortunate it wasn't caught by our automated tests. There seems to be a gap in our unit test coverage. Could you please add a regression test for the above issue as part of this PR?

#44921 also enumerates three other config options which need to be looked into at the same time: `mtu`, `max-concurrent-downloads` and `max-concurrent-uploads`. That issue is not fixed (and therefore this PR is not complete) until validation of those other options are also fixed and the TODO comment removed.
https://github.com/moby/moby/blob/0d1682192282ced1993314e974e5390814fd17ff/daemon/config/config.go#L662-L674

Also, this test can probably be uncommented.
https://github.com/moby/moby/blob/6226988c1e76926593b8408394eee24509b56ef3/daemon/config/config_test.go#L306-L318

psaintlaurent commented:
> This change behaves incorrectly.
> 
> ```
> root@56e9443a7364:/go/src/github.com/docker/docker# cat /etc/docker/daemon.json
> {
> 	"max-download-attempts": 7,
> 	"runtimes": {
> 		"crun": {
> 			"path": "/usr/local/bin/crun"
> 		}
> 	}
> }
> root@56e9443a7364:/go/src/github.com/docker/docker# bundles/dynbinary-daemon/dockerd --userland-proxy=false -D 2>&1| grep 'Max Download Attempts'
> time="2024-06-28T17:59:13.584092212Z" level=debug msg="Max Download Attempts: 5"
> ^C
> ```
> 
> It's unfortunate it wasn't caught by our automated tests. There seems to be a gap in our unit test coverage. Could you please add a regression test for the above issue as part of this PR?
> 
> #44921 also enumerates three other config options which need to be looked into at the same time: `mtu`, `max-concurrent-downloads` and `max-concurrent-uploads`. That issue is not fixed (and therefore this PR is not complete) until validation of those other options are also fixed and the TODO comment removed.
> 
> https://github.com/moby/moby/blob/0d1682192282ced1993314e974e5390814fd17ff/daemon/config/config.go#L662-L674
> 
> Also, this test can probably be uncommented.
> 
> https://github.com/moby/moby/blob/6226988c1e76926593b8408394eee24509b56ef3/daemon/config/config_test.go#L306-L318

I fixed the regression, however the internal user story only addressed max-download-attempts not the rest of the issues in the public Issue list.  Each one may have its own seperate problems that might be best kept as invdividual pull requests for clarity. Can we create new user stories for the other flags and I'll address those separately.  That's probably a dicussion for the internal user story.

corhere commented:
This PR is described as fixing #44921. The internal user story says "Description and A/C in #44921." The acceptance criteria for #44921 are to fix the regression _for all four integer config values_. This PR is therefore incomplete.

psaintlaurent commented:
> This PR is described as fixing #44921. The internal user story says "Description and A/C in #44921." The acceptance criteria for #44921 are to fix the regression _for all four integer config values_. This PR is therefore incomplete.

You asked me to add the PR issue #, that was not the original description from the internal user story.

corhere commented:
<img width="644" alt="Screenshot 2024-06-28 at 6 10 00 PM" src="https://github.com/moby/moby/assets/274527/b3a72b51-edff-44ea-903b-cb1dfe444ee7">

psaintlaurent commented:
> <img alt="Screenshot 2024-06-28 at 6 10 00 PM" width="644" src="https://private-user-images.githubusercontent.com/274527/344326901-b3a72b51-edff-44ea-903b-cb1dfe444ee7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk2MTI5MDYsIm5iZiI6MTcxOTYxMjYwNiwicGF0aCI6Ii8yNzQ1MjcvMzQ0MzI2OTAxLWIzYTcyYjUxLWVkZmYtNDRlYS05MDNiLWNiMWRmZTQ0NGVlNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyOFQyMjEwMDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03YzNiNTMyNzc3ZjVmMjA5MTEyMzQwYmYyNGYyYTNjZWZmMTI5NTlkMmI4MGJlNDk4NTIxZmZlYTI4M2U1YTczJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.mZRpy_o3a29Z6Z8NKRr1Q38QIcsdHxzXT7Z1Q_vRJhs">

I'm aware of th=e description, we discussed changing a single flag as an introduction to docker.  I wasn't supposed to handle the entire issue.  The tasks are only for onboarding and understanding the deve process for moby and docker.
